src/weathergen/datasets/data_reader_synop.py

```diff
@@ -0,0 +1,274 @@
+# (C) Copyright 2025 WeatherGenerator contributors.
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+import logging
+from pathlib import Path
+from typing import override
+
+import numpy as np
+import xarray as xr
+from numpy.typing import NDArray
+
+from weathergen.datasets.data_reader_base import (
+    DataReaderTimestep,
+    ReaderData,
+    TimeWindowHandler,
+    TIndex,
+    check_reader_data,
+)
+
+_logger = logging.getLogger(__name__)
+
+
+class DataReaderSynop(DataReaderTimestep):
+    "Wrapper for SYNOP datasets from MetNo in NetCDF"
+
+    def __init__(
+        self,
+        tw_handler: TimeWindowHandler,
+        filename: Path,
+        stream_info: dict,
+    ) -> None:
+        """
+        Construct data reader for anemoi dataset
+
+        Parameters
+        ----------
+        filename :
+            filename (and path) of dataset
+        stream_info :
+            information about stream
+
+        Returns
+        -------
+        None
+        """
+
+        np32 = np.float32
+
+        # open  dataset to peak that it is compatible with requested parameters
+        ds = xr.open_dataset(filename, engine="netcdf4")
+
+        # If there is no overlap with the time range, the dataset will be empty
+        if tw_handler.t_start >= ds.time.max() or tw_handler.t_end <= ds.time.min():
+            name = stream_info["name"]
+            _logger.warning(f"{name} is not supported over data loader window. Stream is skipped.")
+            super().__init__(tw_handler, stream_info)
+            self._init_empty()
+            return
+
+        if "frequency" in stream_info:
+            assert False, "Frequency sub-sampling currently not supported"
+
+        period = (ds.time[1] - ds.time[0]).values
+        data_start_time = ds.time[0].values
+        data_end_time = ds.time[-1].values
+        assert data_start_time is not None and data_end_time is not None, (
+            data_start_time,
+            data_end_time,
+        )
+        super().__init__(
+            tw_handler,
+            stream_info,
+            data_start_time,
+            data_end_time,
+            period,
+        )
+        # If there is no overlap with the time range, no need to keep the dataset.
+        if tw_handler.t_start >= data_end_time or tw_handler.t_end <= data_start_time:
+            self._init_empty()
+            return
+        else:
+            self.ds = ds
+            self.len = len(ds)
+
+        self.offset_data_channels = 4
+        self.fillvalue = ds["air_temperature"][0, 0].values.item()
```

Comment:
I think we should force the user to set a _FillValue attribute in the NetCDF files if there are missing data. I forgot to do this with my file (well actually, if you are using the default fill value of 9.96921e+36, then you don't need to set the flag, however xarray doesn't recognize this).

When _FillValue is set, xr.open_dataset fill automatically convert missing values to NaNs.